File: server/opensandbox_server/services/factory.py
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Factory for creating sandbox service instances.

This module provides a factory function to create sandbox service implementations
based on application configuration loaded from sandbox_server.config.
"""

import logging
from typing import Optional

from opensandbox_server.config import AppConfig, get_config
from opensandbox_server.services.docker import DockerSandboxService
from opensandbox_server.services.k8s import KubernetesSandboxService
from opensandbox_server.services.fast_sandbox import FastSandboxService
from opensandbox_server.services.sandbox_service import SandboxService
from opensandbox_server.services.composite_service import CompositeSandboxService

logger = logging.getLogger(__name__)


def create_sandbox_service(
    service_type: Optional[str] = None,
    config: Optional[AppConfig] = None,
) -> SandboxService:
    """
    Create a sandbox service instance based on configuration.

    Args:
        service_type: Optional override for service implementation type.
        config: Optional application configuration. Defaults to global config.

    Returns:
        SandboxService: An instance of the configured sandbox service implementation.

    Raises:
        ValueError: If the configured service type is not supported.
    """
    active_config = config or get_config()
    selected_type = (service_type or active_config.runtime.type).lower()

    logger.info(f"Creating sandbox service with type: {selected_type}")

    # Service implementation registry
    # Add new implementations here as they are created
    implementations = {
        "docker": DockerSandboxService,
        "kubernetes": KubernetesSandboxService,
    }

    if selected_type not in implementations:
        supported_types = ", ".join(implementations.keys())
        raise ValueError(
            f"Unsupported sandbox service type: {selected_type}. "
            f"Supported types: {supported_types}"
        )

    if selected_type == "kubernetes":
        implementation = KubernetesSandboxService(config=active_config)
        return CompositeSandboxService(
            implementation,
            FastSandboxService(active_config, k8s_client=implementation.k8s_client),
        )
    return implementations[selected_type](config=active_config)

File: kubernetes/internal/controller/poolassign/factory.go
// Copyright 2025 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package assign

import "fmt"

type weightedScorer struct {
	scorer
	weight int64
}

type predicateCreator func(args map[string]interface{}) (predicate, error)
type scorerCreator func(args map[string]interface{}) (scorer, error)

var (
	predicateRegistry = map[string]predicateCreator{}
	scorerRegistry    = map[string]scorerCreator{}
)

func registerPredicate(name string, creator predicateCreator) {
	predicateRegistry[name] = creator
}

func registerScorer(name string, creator scorerCreator) {
	scorerRegistry[name] = creator
}

func newPredicates(profile *Profile) ([]predicate, error) {
	var predicates []predicate
	for _, name := range profile.Plugins.Predicate {
		creator, ok := predicateRegistry[name]
		if !ok {
			continue
		}
		p, err := creator(pluginArgs(profile, name))
		if err != nil {
			return nil, fmt.Errorf("predicate %q: %w", name, err)
		}
		predicates = append(predicates, p)
	}
	return predicates, nil
}

func newScorers(profile *Profile) ([]weightedScorer, error) {
	var scorers []weightedScorer
	for _, spec := range profile.Plugins.Score {
		creator, ok := scorerRegistry[spec.Name]
		if !ok {
			continue
		}
		s, err := creator(pluginArgs(profile, spec.Name))
		if err != nil {
			return nil, fmt.Errorf("scorer %q: %w", spec.Name, err)
		}
		scorers = append(scorers, weightedScorer{
			scorer: s,
			weight: spec.Weight,
		})
	}
	return scorers, nil
}

func pluginArgs(profile *Profile, pluginName string) map[string]interface{} {
	for _, conf := range profile.PluginConf {
		if conf.Name == pluginName {
			return conf.Args
		}
	}
	return nil
}
